File: graphlib-3/go/lib/paths.go
package graphlib

import (
	"container/heap"
	"math"
	"sort"
)

// PriorityQueueItem represents an item in the priority queue
type PriorityQueueItem struct {
	node     string
	distance float64
	index    int
}

// PriorityQueue implements heap.Interface
type PriorityQueue []*PriorityQueueItem

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].distance < pq[j].distance
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	item := x.(*PriorityQueueItem)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// ShortestPath finds the shortest path between two nodes
func ShortestPath(graphID, startNode, endNode string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{
			"exists":   false,
			"path":     []string{},
			"distance": -1.0,
		}
	}

	if !g.Nodes[startNode] || !g.Nodes[endNode] {
		return map[string]interface{}{
			"exists":   false,
			"path":     []string{},
			"distance": -1.0,
		}
	}

	// Same node
	if startNode == endNode {
		return map[string]interface{}{
			"exists":   true,
			"path":     []string{startNode},
			"distance": 0.0,
		}
	}

	// Use BFS for unweighted graphs, Dijkstra for weighted
	if !g.Weighted {
		return bfsShortestPath(g, startNode, endNode)
	}
	return dijkstraShortestPath(g, startNode, endNode)
}

func bfsShortestPath(g *Graph, startNode, endNode string) map[string]interface{} {
	parent := make(map[string]string)
	visited := make(map[string]bool)

	queue := []string{startNode}
	visited[startNode] = true

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		if node == endNode {
			// Reconstruct path
			path := []string{}
			current := endNode
			for current != "" {
				path = append([]string{current}, path...)
				current = parent[current]
			}
			return map[string]interface{}{
				"exists":   true,
				"path":     path,
				"distance": float64(len(path) - 1),
			}
		}

		// Get sorted neighbors for deterministic order
		neighbors := make([]string, 0, len(g.Adj[node]))
		for neighbor := range g.Adj[node] {
			neighbors = append(neighbors, neighbor)
		}
		sort.Strings(neighbors)

		for _, neighbor := range neighbors {
			if !visited[neighbor] {
				visited[neighbor] = true
				parent[neighbor] = node
				queue = append(queue, neighbor)
			}
		}
	}

	return map[string]interface{}{
		"exists":   false,
		"path":     []string{},
		"distance": -1.0,
	}
}

func dijkstraShortestPath(g *Graph, startNode, endNode string) map[string]interface{} {
	dist := make(map[string]float64)
	parent := make(map[string]string)

	for node := range g.Nodes {
		dist[node] = math.Inf(1)
	}
	dist[startNode] = 0

	pq := &PriorityQueue{}
	heap.Init(pq)
	heap.Push(pq, &PriorityQueueItem{node: startNode, distance: 0})

	visited := make(map[string]bool)

	for pq.Len() > 0 {
		item := heap.Pop(pq).(*PriorityQueueItem)
		node := item.node

		if visited[node] {
			continue
		}
		visited[node] = true

		if node == endNode {
			// Reconstruct path
			path := []string{}
			current := endNode
			for current != "" {
				path = append([]string{current}, path...)
				current = parent[current]
			}
			return map[string]interface{}{
				"exists":   true,
				"path":     path,
				"distance": dist[endNode],
			}
		}

		for neighbor, weight := range g.Adj[node] {
			if !visited[neighbor] {
				newDist := dist[node] + weight
				if newDist < dist[neighbor] {
					dist[neighbor] = newDist
					parent[neighbor] = node
					heap.Push(pq, &PriorityQueueItem{node: neighbor, distance: newDist})
				}
			}
		}
	}

	return map[string]interface{}{
		"exists":   false,
		"path":     []string{},
		"distance": -1.0,
	}
}

// AllShortestPaths finds shortest paths from start node to all other nodes
func AllShortestPaths(graphID, startNode string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{
			"distances":   map[string]float64{},
			"paths":       map[string][]string{},
			"unreachable": []string{},
		}
	}

	if !g.Nodes[startNode] {
		return map[string]interface{}{
			"distances":   map[string]float64{},
			"paths":       map[string][]string{},
			"unreachable": []string{},
		}
	}

	dist := make(map[string]float64)
	parent := make(map[string]string)

	for node := range g.Nodes {
		dist[node] = math.Inf(1)
	}
	dist[startNode] = 0

	pq := &PriorityQueue{}
	heap.Init(pq)
	heap.Push(pq, &PriorityQueueItem{node: startNode, distance: 0})

	visited := make(map[string]bool)

	for pq.Len() > 0 {
		item := heap.Pop(pq).(*PriorityQueueItem)
		node := item.node

		if visited[node] {
			continue
		}
		visited[node] = true

		for neighbor, weight := range g.Adj[node] {
			w := weight
			if !g.Weighted {
				w = 1.0
			}
			if !visited[neighbor] {
				newDist := dist[node] + w
				if newDist < dist[neighbor] {
					dist[neighbor] = newDist
					parent[neighbor] = node
					heap.Push(pq, &PriorityQueueItem{node: neighbor, distance: newDist})
				}
			}
		}
	}

	// Build result
	distances := make(map[string]float64)
	paths := make(map[string][]string)
	unreachable := []string{}

	for node := range g.Nodes {
		if math.IsInf(dist[node], 1) {
			unreachable = append(unreachable, node)
		} else {
			distances[node] = dist[node]
			// Reconstruct path
			path := []string{}
			current := node
			for current != "" {
				path = append([]string{current}, path...)
				current = parent[current]
			}
			paths[node] = path
		}
	}

	sort.Strings(unreachable)

	return map[string]interface{}{
		"distances":   distances,
		"paths":       paths,
		"unreachable": unreachable,
	}
}

// HasPath checks if a path exists between two nodes
func HasPath(graphID, startNode, endNode string) map[string]interface{} {
	g, ok := Store.GetGraph(graphID)
	if !ok {
		return map[string]interface{}{"exists": false}
	}

	if !g.Nodes[startNode] || !g.Nodes[endNode] {
		return map[string]interface{}{"exists": false}
	}

	// Same node
	if startNode == endNode {
		return map[string]interface{}{"exists": true}
	}

	// BFS to check reachability
	visited := make(map[string]bool)
	queue := []string{startNode}
	visited[startNode] = true

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		for neighbor := range g.Adj[node] {
			if neighbor == endNode {
				return map[string]interface{}{"exists": true}
			}
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return map[string]interface{}{"exists": false}
}

File: graphlib-3/ts/src/paths.ts
import {
  ShortestPathResult,
  AllShortestPathsResult,
  HasPathResult,
} from './types';
import { getGraph } from './graph';

export function shortestPath(
  graphId: string,
  startNode: string,
  endNode: string
): ShortestPathResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false, path: [], distance: -1 };
  }

  if (!graph.nodes.has(startNode) || !graph.nodes.has(endNode)) {
    return { exists: false, path: [], distance: -1 };
  }

  if (startNode === endNode) {
    return { exists: true, path: [startNode], distance: 0 };
  }

  // Use Dijkstra for weighted, BFS for unweighted
  if (graph.weighted) {
    return dijkstra(graphId, startNode, endNode);
  } else {
    return bfsPath(graphId, startNode, endNode);
  }
}

function bfsPath(
  graphId: string,
  startNode: string,
  endNode: string
): ShortestPathResult {
  const graph = getGraph(graphId)!;
  const parent = new Map<string, string>();
  const visited = new Set<string>();
  const queue: string[] = [startNode];
  visited.add(startNode);

  while (queue.length > 0) {
    const current = queue.shift()!;

    if (current === endNode) {
      // Reconstruct path
      const path: string[] = [];
      let node: string | undefined = endNode;
      while (node) {
        path.unshift(node);
        node = parent.get(node);
      }
      return { exists: true, path, distance: path.length - 1 };
    }

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      const sortedNeighbors = Array.from(neighbors.keys()).sort();
      for (const neighbor of sortedNeighbors) {
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          parent.set(neighbor, current);
          queue.push(neighbor);
        }
      }
    }
  }

  return { exists: false, path: [], distance: -1 };
}

function dijkstra(
  graphId: string,
  startNode: string,
  endNode: string
): ShortestPathResult {
  const graph = getGraph(graphId)!;
  const distances = new Map<string, number>();
  const parent = new Map<string, string>();
  const visited = new Set<string>();

  for (const node of graph.nodes) {
    distances.set(node, Infinity);
  }
  distances.set(startNode, 0);

  // Simple priority queue using array (adequate for correctness)
  const pq: Array<[number, string]> = [[0, startNode]];

  while (pq.length > 0) {
    // Find minimum
    pq.sort((a, b) => a[0] - b[0]);
    const [dist, current] = pq.shift()!;

    if (visited.has(current)) continue;
    visited.add(current);

    if (current === endNode) {
      const path: string[] = [];
      let node: string | undefined = endNode;
      while (node) {
        path.unshift(node);
        node = parent.get(node);
      }
      return { exists: true, path, distance: dist };
    }

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const [neighbor, weight] of neighbors) {
        if (!visited.has(neighbor)) {
          const newDist = dist + weight;
          if (newDist < distances.get(neighbor)!) {
            distances.set(neighbor, newDist);
            parent.set(neighbor, current);
            pq.push([newDist, neighbor]);
          }
        }
      }
    }
  }

  return { exists: false, path: [], distance: -1 };
}

export function allShortestPaths(
  graphId: string,
  startNode: string
): AllShortestPathsResult {
  const graph = getGraph(graphId);
  if (!graph || !graph.nodes.has(startNode)) {
    return { distances: {}, paths: {}, unreachable: [] };
  }

  const distances: Record<string, number> = {};
  const paths: Record<string, string[]> = {};
  const parent = new Map<string, string>();
  const distMap = new Map<string, number>();
  const visited = new Set<string>();

  for (const node of graph.nodes) {
    distMap.set(node, Infinity);
  }
  distMap.set(startNode, 0);

  if (graph.weighted) {
    // Dijkstra
    const pq: Array<[number, string]> = [[0, startNode]];

    while (pq.length > 0) {
      pq.sort((a, b) => a[0] - b[0]);
      const [dist, current] = pq.shift()!;

      if (visited.has(current)) continue;
      visited.add(current);

      const neighbors = graph.adjacency.get(current);
      if (neighbors) {
        for (const [neighbor, weight] of neighbors) {
          if (!visited.has(neighbor)) {
            const newDist = dist + weight;
            if (newDist < distMap.get(neighbor)!) {
              distMap.set(neighbor, newDist);
              parent.set(neighbor, current);
              pq.push([newDist, neighbor]);
            }
          }
        }
      }
    }
  } else {
    // BFS
    const queue: string[] = [startNode];
    visited.add(startNode);

    while (queue.length > 0) {
      const current = queue.shift()!;
      const currentDist = distMap.get(current)!;

      const neighbors = graph.adjacency.get(current);
      if (neighbors) {
        const sortedNeighbors = Array.from(neighbors.keys()).sort();
        for (const neighbor of sortedNeighbors) {
          if (!visited.has(neighbor)) {
            visited.add(neighbor);
            distMap.set(neighbor, currentDist + 1);
            parent.set(neighbor, current);
            queue.push(neighbor);
          }
        }
      }
    }
  }

  const unreachable: string[] = [];

  for (const node of graph.nodes) {
    const dist = distMap.get(node)!;
    if (dist === Infinity) {
      unreachable.push(node);
    } else {
      distances[node] = dist;
      // Reconstruct path
      const path: string[] = [];
      let current: string | undefined = node;
      while (current) {
        path.unshift(current);
        current = parent.get(current);
      }
      paths[node] = path;
    }
  }

  return { distances, paths, unreachable: unreachable.sort() };
}

export function hasPath(
  graphId: string,
  startNode: string,
  endNode: string
): HasPathResult {
  const graph = getGraph(graphId);
  if (!graph) {
    return { exists: false };
  }

  if (!graph.nodes.has(startNode) || !graph.nodes.has(endNode)) {
    return { exists: false };
  }

  if (startNode === endNode) {
    return { exists: true };
  }

  // BFS to check reachability
  const visited = new Set<string>();
  const queue: string[] = [startNode];
  visited.add(startNode);

  while (queue.length > 0) {
    const current = queue.shift()!;

    const neighbors = graph.adjacency.get(current);
    if (neighbors) {
      for (const neighbor of neighbors.keys()) {
        if (neighbor === endNode) {
          return { exists: true };
        }
        if (!visited.has(neighbor)) {
          visited.add(neighbor);
          queue.push(neighbor);
        }
      }
    }
  }

  return { exists: false };
}
